### PR TITLE
Fix stake redelegate

### PR DESCRIPTION
**Pull request by t-nelson**
#### Problem

Re-delegating a formerly delegated, but currently deactivated stake account, does not consider changes to the account balance

#### Summary of Changes

Use the current account balance rather than the previous delegation amount

WIP until activation is coordinated with #13357

### CODE REVIEW

**Comment by t-nelson** [comment]
Do you have time to take a peek at one more, @rwalker-com? :pray:

File: programs/stake/src/stake_state.rs [hunk 1]
```diff
@@ -3689,4 +3692,86 @@ mod tests {
...
+        let vote_pubkey = Pubkey::new_unique();
+        let vote_account = RefCell::new(vote_state::create_account(
+            &vote_pubkey,
+            &Pubkey::new_unique(),
+            0,
+            100,
+        ));
+        let vote_keyed_account = KeyedAccount::new(&vote_pubkey, false, &vote_account);
+
+        let signers: HashSet<_> = [withdrawer_pubkey].iter().cloned().collect();
```
Comment (rwalker-apple) [inline programs/stake/src/stake_state.rs:3726]: nit: is this the tightest way to construct this thingie?

**Review by rwalker-apple — APPROVED** [review]

File: programs/stake/src/stake_state.rs (hunk 1 above)
Comment (rwalker-com) [inline programs/stake/src/stake_state.rs:3726]: nit: is this the tightest way to construct this thingie?

File: programs/stake/src/stake_state.rs [hunk 2]
```diff
@@ -3689,4 +3692,86 @@ mod tests {
         // Test another staking action
         assert_eq!(stake_keyed_account.deactivate(&clock, &new_signers), Ok(()));
     }
+
+    #[test]
+    fn test_redelegate_considers_withdraws() {
```
Comment (rwalker-com) [inline programs/stake/src/stake_state.rs:3697]: nice name.  covering an increase in account balance might reassure.

**Review by rwalker-com — APPROVED** [review]
great job closing a hole you can throw a cat through

File: programs/stake/src/stake_state.rs (hunk 1 above)
Comment (t-nelson) [reply programs/stake/src/stake_state.rs:3726]: Definitely not. Thanks for the prod! I meant to fix that up

File: programs/stake/src/stake_state.rs (hunk 2 above)
Comment (t-nelson) [reply programs/stake/src/stake_state.rs:3697]: Good point!  I'll add that, and make the name even better :wink:

**Review by CriesofCarrots — APPROVED** [review]
I'll take another look when you push your test changes, but fix lgtm!
It'll be good to get another winky comment in the code base as well ;)

**Comment by t-nelson** [comment]
Review's been addressed!

**Comment by CriesofCarrots** [comment]
This will need rebasing on #13394 (make sure it goes in stake_state.rs, as opposed to the legacy mod)

**Comment by t-nelson** [comment]
Rebased on #13394

**Pull request merged**